Escreva uma função que receba três números e retorne o maior entre eles. Não use a função max().
➤ Reforce o uso de condicionais aninhadas.

In [ ]:
maior = 0
menor = 0

for i in range(3):
  num = int(input('Digite um número: '))
  if num > maior:
    maior = num
  elif num < maior:
    menor = num

print(f'O maior número digitado foi {maior}')
print(f'O menor número digitado foi {menor}')

Digite um número: 5
Digite um número: 9
Digite um número: 3
O maior número digitado foi 9
O menor número digitado foi 3


Crie uma função que determine se um número é par ou ímpar e, adicionalmente, se ele é positivo, negativo ou nulo.

In [ ]:
num = int(input('Digite um número: '))

if num % 2 == 0:
  print(f'O número {num} é par')
else:
  print(f'O número {num} é impar')

if num > 0:
  print(f'O número informado {num} é um número positivo')
elif num < 0:
  print(f'O número informado {num} é um número negativo')
else:
  print(f'O número informado {num} é um número nulo')

Digite um número: -150
O número -150 é par
O número informado -150 é um número negativo


Faça uma função que receba a idade e o tempo de contribuição de uma pessoa, e diga se ela pode se aposentar pelas regras:

Homens: 35 anos de contribuição e idade mínima de 60 anos

Mulheres: 30 anos de contribuição e idade mínima de 55 anos

In [ ]:
nome = input('Informe o seu nome: ')
while genero != 'M' and genero != 'F':
  genero = input('\nInforme o seu gênero(M,F): ')
idade = int(input('\nInforme a sua idade: '))
tempo_contribuicao = int(input('\nInforme o tempo de contribuição(Anos):'))


if genero == 'M' and tempo_contribuicao >= 35 and idade >= 60:
  print(f'\nParabéns {nome}! Você já pode se aposentar.')
elif genero == 'F' and tempo_contribuicao >= 30 and idade >= 55:
  print(f'\nParabéns {nome}! Você pode se aposentar.')
else:
  print(f'\nInfelizmente {nome}, Você não pode se aposentar...')

Informe o seu nome: William

Informe o seu gênero(M,F): C

Informe o seu gênero(M,F): B

Informe o seu gênero(M,F): D

Informe o seu gênero(M,F): K

Informe o seu gênero(M,F): ,

Informe o seu gênero(M,F): M

Informe a sua idade: 65

Informe o tempo de contribuição(Anos):30

Infelizmente William, Você não pode se aposentar...


Implemente uma função que receba um número inteiro n e retorne uma lista com os n primeiros números primos.

- num = 2  # primeiro número candidato

- num ** 0.5

Eleva num a 0.5, o que é o mesmo que tirar a raiz quadrada.

Por exemplo: 9 ** 0.5 → 3.0.

Por que só até a raiz quadrada?

Se num não tiver nenhum divisor menor ou igual à sua raiz quadrada, ele não terá divisores maiores (exceto ele mesmo).

Exemplo: se num = 36, a raiz é 6.
Se não acharmos divisor até 6, significa que não haverá nenhum divisor depois disso (além do próprio 36).

Isso reduz muito a quantidade de testes, deixando o código mais rápido.

O + 1

O range() em Python vai até o valor especificado, mas não o inclui.

Como queremos incluir o valor da raiz quadrada no teste, somamos + 1.

In [ ]:
print('Vou te mostrar uma lista de números primos.')

n = int(input("Digite a quantidade de números primos desejada: "))

primos = []
# Inicializa a variável num com o valor 2. num representa o número atual que está sendo verificado para saber se é primo. Começamos com 2, pois é o primeiro número primo.
num = 2

while len(primos) < n:
  for i in range(2, int(num ** 0.5) + 1): # vazio
      if num % i == 0:
          break
  else:
      primos.append(num)
  num += 1

print(primos)

Vou te mostrar uma lista de números primos.
Digite a quantidade de números primos desejada: 25
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


Vamos simular a execução do código para encontrar os 5 primeiros números primos (`n = 5`).

1.  `print('Vou te mostrar uma lista de números primos.')`
    *   **Explicação:** Esta linha simplesmente imprime uma mensagem para o usuário.
    *   **Saída:** `Vou te mostrar uma lista de números primos.`

2.  `n = int(input("Digite a quantidade de números primos desejada: "))`
    *   **Explicação:** Pede ao usuário para digitar um número e armazena esse valor na variável `n` como um inteiro.
    *   **Simulação:** O usuário digita `5`.
    *   **Estado das variáveis:** `n = 5`

3.  `primos = []`
    *   **Explicação:** Inicializa uma lista vazia chamada `primos` para armazenar os números primos encontrados.
    *   **Estado das variáveis:** `primos = []`

4.  `num = 2`
    *   **Explicação:** Inicializa a variável `num` com o valor 2. `num` representa o número atual que está sendo verificado para saber se é primo. Começamos com 2, pois é o primeiro número primo.
    *   **Estado das variáveis:** `num = 2`

5.  `while len(primos) < n:`
    *   **Explicação:** Este é o loop principal. Ele continua executando enquanto o número de elementos na lista `primos` for menor que `n` (que é 5 na nossa simulação).
    *   **Simulação (Início da primeira iteração):** `len(primos)` é 0, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Explicação:** Este é o loop interno. Ele verifica se `num` é divisível por qualquer número `i` no intervalo de 2 até a raiz quadrada de `num`, incluindo a raiz quadrada. Se encontrar um divisor, `num` não é primo.
    *   **Simulação (Primeira iteração):** `num` é 2. `int(num ** 0.5) + 1` é `int(2 ** 0.5) + 1` = `int(1.414...) + 1` = `1 + 1` = 2. O `range(2, 2)` está vazio. O loop `for` não executa nenhuma iteração.

7.  `else:`
    *   **Explicação:** Este bloco `else` pertence ao loop `for`. Ele é executado APENAS se o loop `for` completar todas as suas iterações sem encontrar um `break`. Isso significa que nenhum divisor foi encontrado, e `num` é primo.
    *   **Simulação (Primeira iteração):** O loop `for` não executou, então ele completou sem um `break`. O bloco `else` é executado.

8.  `primos.append(num)`
    *   **Explicação:** Adiciona o número `num` (que foi determinado como primo) à lista `primos`.
    *   **Simulação (Primeira iteração):** `primos.append(2)`.
    *   **Estado das variáveis:** `primos = [2]`

9.  `num += 1`
    *   **Explicação:** Incrementa `num` em 1 para verificar o próximo número.
    *   **Simulação (Primeira iteração):** `num` se torna 3.
    *   **Estado das variáveis:** `num = 3`

---

**Início da Segunda Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 1, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 3. `int(num ** 0.5) + 1` é `int(3 ** 0.5) + 1` = `int(1.732...) + 1` = `1 + 1` = 2. O `range(2, 2)` está vazio. O loop `for` não executa nenhuma iteração.

7.  `else:`
    *   **Simulação:** O loop `for` completou sem um `break`. O bloco `else` é executado.

8.  `primos.append(num)`
    *   **Simulação:** `primos.append(3)`.
    *   **Estado das variáveis:** `primos = [2, 3]`

9.  `num += 1`
    *   **Simulação:** `num` se torna 4.
    *   **Estado das variáveis:** `num = 4`

---

**Início da Terceira Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 2, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 4. `int(num ** 0.5) + 1` é `int(4 ** 0.5) + 1` = `int(2.0) + 1` = `2 + 1` = 3. O `range(2, 3)` irá gerar `i = 2`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `4 % 2 == 0` é verdadeiro.
        *   `break`
        *   **Simulação:** O loop `for` é interrompido.

7.  `else:`
    *   **Simulação:** O bloco `else` NÃO é executado porque o loop `for` foi interrompido por um `break`.

9.  `num += 1`
    *   **Simulação:** `num` se torna 5.
    *   **Estado das variáveis:** `num = 5`

---

**Início da Quarta Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 2, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 5. `int(num ** 0.5) + 1` é `int(5 ** 0.5) + 1` = `int(2.236...) + 1` = `2 + 1` = 3. O `range(2, 3)` irá gerar `i = 2`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `5 % 2 == 0` é falso. A condição não é atendida.
    *   **Fim da iteração do `for` com `i = 2`.** O loop `for` terminou suas iterações.

7.  `else:`
    *   **Simulação:** O loop `for` completou sem um `break`. O bloco `else` é executado.

8.  `primos.append(num)`
    *   **Simulação:** `primos.append(5)`.
    *   **Estado das variáveis:** `primos = [2, 3, 5]`

9.  `num += 1`
    *   **Simulação:** `num` se torna 6.
    *   **Estado das variáveis:** `num = 6`

---

**Início da Quinta Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 3, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 6. `int(num ** 0.5) + 1` é `int(6 ** 0.5) + 1` = `int(2.449...) + 1` = `2 + 1` = 3. O `range(2, 3)` irá gerar `i = 2`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `6 % 2 == 0` é verdadeiro.
        *   `break`
        *   **Simulação:** O loop `for` é interrompido.

7.  `else:`
    *   **Simulação:** O bloco `else` NÃO é executado porque o loop `for` foi interrompido por um `break`.

9.  `num += 1`
    *   **Simulação:** `num` se torna 7.
    *   **Estado das variáveis:** `num = 7`

---

**Início da Sexta Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 3, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 7. `int(num ** 0.5) + 1` é `int(7 ** 0.5) + 1` = `int(2.645...) + 1` = `2 + 1` = 3. O `range(2, 3)` irá gerar `i = 2`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `7 % 2 == 0` é falso. A condição não é atendida.
    *   **Fim da iteração do `for` com `i = 2`.** O loop `for` terminou suas iterações.

7.  `else:`
    *   **Simulação:** O loop `for` completou sem um `break`. O bloco `else` é executado.

8.  `primos.append(num)`
    *   **Simulação:** `primos.append(7)`.
    *   **Estado das variáveis:** `primos = [2, 3, 5, 7]`

9.  `num += 1`
    *   **Simulação:** `num` se torna 8.
    *   **Estado das variáveis:** `num = 8`

---

**Início da Sétima Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 4, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 8. `int(num ** 0.5) + 1` é `int(8 ** 0.5) + 1` = `int(2.828...) + 1` = `2 + 1` = 3. O `range(2, 3)` irá gerar `i = 2`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `8 % 2 == 0` é verdadeiro.
        *   `break`
        *   **Simulação:** O loop `for` é interrompido.

7.  `else:`
    *   **Simulação:** O bloco `else` NÃO é executado porque o loop `for` foi interrompido por um `break`.

9.  `num += 1`
    *   **Simulação:** `num` se torna 9.
    *   **Estado das variáveis:** `num = 9`

---

**Início da Oitava Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 4, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 9. `int(num ** 0.5) + 1` é `int(9 ** 0.5) + 1` = `int(3.0) + 1` = `3 + 1` = 4. O `range(2, 4)` irá gerar `i = 2` e `i = 3`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `9 % 2 == 0` é falso. A condição não é atendida.
    *   **Início da iteração do `for` com `i = 3`:**
        *   `if num % i == 0:`
        *   **Simulação:** `9 % 3 == 0` é verdadeiro.
        *   `break`
        *   **Simulação:** O loop `for` é interrompido.

7.  `else:`
    *   **Simulação:** O bloco `else` NÃO é executado porque o loop `for` foi interrompido por um `break`.

9.  `num += 1`
    *   **Simulação:** `num` se torna 10.
    *   **Estado das variáveis:** `num = 10`

---

**Início da Nona Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 4, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 10. `int(num ** 0.5) + 1` é `int(10 ** 0.5) + 1` = `int(3.162...) + 1` = `3 + 1` = 4. O `range(2, 4)` irá gerar `i = 2` e `i = 3`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `10 % 2 == 0` é verdadeiro.
        *   `break`
        *   **Simulação:** O loop `for` é interrompido.

7.  `else:`
    *   **Simulação:** O bloco `else` NÃO é executado porque o loop `for` foi interrompido por um `break`.

9.  `num += 1`
    *   **Simulação:** `num` se torna 11.
    *   **Estado das variáveis:** `num = 11`

---

**Início da Décima Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 4, que é menor que 5. O loop continua.

6.  `for i in range(2, int(num ** 0.5) + 1):`
    *   **Simulação:** `num` é 11. `int(num ** 0.5) + 1` é `int(11 ** 0.5) + 1` = `int(3.316...) + 1` = `3 + 1` = 4. O `range(2, 4)` irá gerar `i = 2` e `i = 3`.
    *   **Início da iteração do `for` com `i = 2`:**
        *   `if num % i == 0:`
        *   **Simulação:** `11 % 2 == 0` é falso. A condição não é atendida.
    *   **Início da iteração do `for` com `i = 3`:**
        *   `if num % i == 0:`
        *   **Simulação:** `11 % 3 == 0` é falso. A condição não é atendida.
    *   **Fim das iterações do `for`.**

7.  `else:`
    *   **Simulação:** O loop `for` completou sem um `break`. O bloco `else` é executado.

8.  `primos.append(num)`
    *   **Simulação:** `primos.append(11)`.
    *   **Estado das variáveis:** `primos = [2, 3, 5, 7, 11]`

9.  `num += 1`
    *   **Simulação:** `num` se torna 12.
    *   **Estado das variáveis:** `num = 12`

---

**Início da Décima Primeira Iteração do `while`:**

5.  `while len(primos) < n:`
    *   **Simulação:** `len(primos)` é 5, que NÃO é menor que 5. A condição do loop `while` é falsa. O loop `while` termina.

10. `print(primos)`
    *   **Explicação:** Imprime a lista final de números primos.
    *   **Simulação:** Imprime o conteúdo da lista `primos`.
    *   **Saída:** `[2, 3, 5, 7, 11]`

Espero que esta explicação detalhada, seguindo o fluxo de execução e mostrando o estado das variáveis, tenha ajudado a esclarecer como o código funciona e como os loops e a decisão do `else` associado ao `for` se comportam.

Se tiver mais alguma dúvida ou quiser simular com outro valor de `n`, é só me dizer!